shared/src/server/health.rs

```diff
@@ -55,3 +71,132 @@ impl std::fmt::Display for HealthCheckStatus {
         }
     }
 }
+
+pub trait HealthCheck {
+    fn status_code(&self) -> u16;
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum ControlPlaneState {
+    Draining,
+    Ok,
+}
+
+impl HealthCheck for ControlPlaneState {
+    fn status_code(&self) -> u16 {
+        match self {
+            ControlPlaneState::Ok => 200,
+            ControlPlaneState::Draining => 500,
+        }
+    }
+}
+
+impl HealthCheck for DataPlaneState {
+    fn status_code(&self) -> u16 {
+        match self {
+            DataPlaneState::Initialized(diagnostic) if diagnostic.is_healthy() => 200,
+            _ => 500,
+        }
+    }
+}
+
+impl HealthCheck for HealthCheckLog {
+    fn status_code(&self) -> u16 {
+        self.status_code()
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum DataPlaneState {
+    Error(String),
+    Unknown(String),
+    Provisioning,
+    Attesting,
+    SourcingTlsCerts,
+    Initialized(DataPlaneDiagnostic),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct DataPlaneDiagnostic {
+    pub user_process: UserProcessHealth,
+}
+
+impl DataPlaneDiagnostic {
+    pub fn is_healthy(&self) -> bool {
+        match self.user_process {
+            UserProcessHealth::Unknown(_) => true,
+            UserProcessHealth::Error(_) => false,
+            UserProcessHealth::Response { status_code, .. } => status_code == 200,
```

Comment:
do we always want to spin the enclave when the UP reports non 200 codes? (Also this should probably be > 300 not == 200)

Reply:
yeah this should be `ok` responses so >= 200, <300